devimint/src/tests.rs

```diff
@@ -350,6 +350,151 @@ pub async fn latency_tests(dev_fed: DevFed, r#type: LatencyTest) -> Result<()> {
     Ok(())
 }
 
+pub async fn upgrade_tests(process_mgr: &ProcessManager, binary: UpgradeTest) -> Result<()> {
+    match binary {
+        UpgradeTest::Fedimintd { mut paths } => {
+            anyhow::ensure!(!paths.is_empty(), "Must provide at least 1 binary path");
+            let oldest_fedimintd = paths.first().expect("at least 1 binary was provided");
+            std::env::set_var("FM_FEDIMINTD_BASE_EXECUTABLE", oldest_fedimintd);
+            let fedimintd_version = crate::util::FedimintdCmd::version_or_default().await;
+
+            info!(
+                "running first stress test for fedimintd version: {}",
+                fedimintd_version
+            );
+            let mut dev_fed = dev_fed(process_mgr).await?;
+            let client = dev_fed.fed.new_joined_client("test-client").await?;
+            try_join!(stress_test_fed(&dev_fed), wait_session(&client))?;
+
+            // add the current fedimintd binary from PATH
+            paths.push("fedimintd".into());
+            for path in paths.iter().skip(1) {
+                dev_fed.fed.restart_all_with_bin(process_mgr, path).await?;
```

Comment:
this doesn't have the staggered shutdown like I had in initial upgrade tests

it is like to have some activity on some peers when upgrading

Reply:
I updated to stagger the shutdown/restart of all peers since that's likely a closer approximation for upgrades in the wild. Let me know if there's a reason to prefer only staggering the last peer.

Reply:
I delayed last peer so other peers could complete a session without